### PR TITLE
OperationOutcome as Error if returned by the FHIR service

### PR DESCRIPTION
**Pull request by tader**
[INT-255: Parse OperationOutcome on the client side](https://linear.app/zorg-bij-jou/issue/INT-255/orca-or-return-operationoutcome-on-error-and-parse-them-client-side)

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -189,12 +189,19 @@ func (d BaseClient) doRequest(httpRequest *http.Request, target any, opts ...Opt
 		if d.config.Non2xxStatusHandler != nil {
 			d.config.Non2xxStatusHandler(httpResponse, data)
 		}
+		// Checking for OperationOutcome
```
Comment (dwelman) [inline client.go:192]: Nitpick: This comment doesn't add any information not provided by the method name in the next line

**Review by dwelman — APPROVED** [review]
Approved, just minor nitpick about comments

File: client.go [hunk 2]
```diff
@@ -189,12 +189,17 @@ func (d BaseClient) doRequest(httpRequest *http.Request, target any, opts ...Opt
...
+		if err = checkForOperationOutcome(data); err != nil {
+			return err
+		}
 		return fmt.Errorf("FHIR request failed (%s %s, status=%d)", httpRequest.Method, httpRequest.URL.String(), httpResponse.StatusCode)
 	}
 	if len(data) > d.config.MaxResponseSize {
 		return fmt.Errorf("FHIR response exceeds max. safety limit of %d bytes (%s %s, status=%d)", d.config.MaxResponseSize, httpRequest.Method, httpRequest.URL.String(), httpResponse.StatusCode)
 	}
-	// TODO: Handle errornous responses (OperationOutcome?)
+	if err = checkForOperationOutcome(data); err != nil {
```
Comment (reinkrul) [inline client.go:200]: > Operation outcomes returned SHOULD be in alignment with the HTTP response code. For example, if the HTTP code indicates a failure (300+), at least one of the issues should have a severity of "error", indicating the reason for the failure.
(https://www.hl7.org/fhir/r4b/operationoutcome.html#using)

and

> Operation outcomes are sets of error, warning and information messages that provide detailed information about the outcome of an attempted system operation. They are provided as a direct system response, or component of one, and provide information about the outcome of the operation.

(note "informational")

In other words, an operation outcome for an OK status could be considered as the opertion succeeded? I think the response the FHIR server returns when creating a Bundle (FHIR transaction) also contains an OperationOutcome.

File: operation_outcome.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+		return false
+	}
+
+	return strings.EqualFold(*r.ResourceType, "OperationOutcome")
+}
+
+func (r OperationOutcome) Error() string {
+	var messages []string
+	for _, issue := range r.Issue {
+		if issue.Diagnostics != nil {
```
Comment (reinkrul) [inline operation_outcome.go:26]: diagnostics being an optional field (and details as well), don't we always want an entry for the particular operationoutcome? Even when diagnostics is not set? (and maybe support detail field as well). Or add a note/TODO that we currently don't support it ehre?

**Review by reinkrul — CHANGES_REQUESTED** [review]

**Comment by tader** [comment]
Hey @reinkrul I have made changes to the PR, I hope I correctly understood you remarks 😉

**Pull request merged**